src/main/java/com/ait/lienzo/client/core/shape/TextTruncateWrapper.java

```diff
@@ -0,0 +1,239 @@
+    }
+
+    public BoundingBox getWrapBoundaries() {
+        return wrapBoundaries;
+    }
+
+    public TextTruncateWrapper setWrapBoundaries(final BoundingBox boundaries) {
+        wrapBoundaries = boundaries;
+        return this;
+    }
+
+    @Override
+    public BoundingBox getBoundingBox() {
+        final double[] boundaries = calculateWrapBoundaries();
+        return new BoundingBox().addX(0).addX(wrapBoundaries.getWidth()).addY(0).addY(boundaries[1]);
+    }
+
+    private double getWrapBoundariesWidth() {
+        final double margin = 4 * getBoundingBoxForString(" ").getWidth();
+        return this.wrapBoundaries.getWidth() - margin;
+    }
+
+    private double[] calculateWrapBoundaries() {
+        final String[] words = textSupplier.get().split("\\s");
+        if (words.length < 1) {
+            return new double[]{wrapBoundaries.getX(), wrapBoundaries.getY()};
+        }
+
+        final double wrapWidth = getWrapBoundariesWidth();
+        final String firstWord = words[0];
+        double width = getBoundingBoxForString(firstWord).getWidth();
+        final StringBuilder nextLine = new StringBuilder(firstWord);
+        int numOfLines = 1;
+        for (int i = 1; i < words.length; i++) {
+            width = getBoundingBoxForString(nextLine + " " + words[i]).getWidth();
+            if (width <= wrapWidth) {
+                nextLine.append(" ").append(words[i]);
+            } else {
+                nextLine.setLength(words[i].length());
+                nextLine.replace(0,
+                                 words[i].length(),
+                                 words[i]);
+                numOfLines++;
+            }
+        }
+
+        final double lineHeight = (getBoundingBoxForString(" ").getHeight());
+
+        while (!hasVerticalSpace(numOfLines, lineHeight, wrapBoundaries.getHeight() - (Y_OFFSET * numOfLines))
+                && numOfLines >= 0) {
+            numOfLines--;
+        }
+
+        final double height = lineHeight * numOfLines;
+
+        return new double[]{width, height};
+    }
+
+    private boolean hasVerticalSpace(final int lineIndex,
+                                     final double lineHeight,
+                                     final double availableHeight) {
+        return lineHeight * (lineIndex + Y_OFFSET) <= availableHeight;
+    }
+
+    @Override
+    public void drawString(final Context2D context,
+                           final Attributes attr,
+                           final IDrawString drawCommand) {
+        final String[] words = attr.getText().split("\\s");
+
+        if (words.length < 1) {
+            return;
+        }
+
+        final double lineHeight = getBoundingBoxForString(" ").getHeight();
+        final ArrayList<String> lines = new ArrayList<String>();
+        final double boundariesWidth = getWrapBoundariesWidth();
+        StringBuilder currentLine = new StringBuilder();
+        String currentWord;
+
+        for (int i = 0; i < words.length; i++) {
```

Comment:
So what does this do that's different to the above?

And if it is doing something different then can the calculation above (in `calculateWrapBoundaries()`) be relied on?

Reply:
This is the precise algorithm considering each edge. It also concatenates words in the same line if this is possible.  The other one is just to create bounding box, ignoring the precision of the width because in fact we never overflows width and use the node width. But for the height we may overflow in some scenarios. 

For this precise algorithm, there is some whiles that edge cases. For example, consider: "Some looooooooooooooooooooong description".
The second word overflows horizontal, so we have to split that word. But maybe when we split that word, we don't have more VERTICAL space available, so instead of split we have to truncate, append "...." and stop.
Or we can have cases like "Short1 short2". Consider that we only have space to 8 chars. So instead of split the word "short2", doing something like:
"short1 s
hort2"
Since we have space for 2 lines, ee improve readability by doing something like:
"short1
short2"

That kind of precise calculus is doing in this method while the other one is just a "worst case estimative" used to create the bounding box of the node without overflowing.